Restructuring vector search docs and adding vector landing page and use cases page (#5567)
elastic/docs-content: solutions/search/vector/vector-search-use-cases.md

To understand how semantic search works in {{es}}, refer to [Semantic search](../semantic-search.md).

:::::

:::::{step} Choose an implementation approach

To implement semantic search, you can choose between the following approaches depending on your level of control and complexity:

- **[Semantic search with `semantic_text`](../semantic-search/semantic-search-semantic-text.md)**  
  Use this for the simplest setup. {{es}} handles chunking, embedding generation, and indexing automatically. Suitable for most standard text search use cases.

- **[Hybrid search with `semantic_text`](../hybrid-semantic-text.md)**  
  Use this when you want to combine semantic understanding with keyword matching. This is recommended for most production use cases where both exact matches and meaning matter.

- **[Semantic search with the {{infer}} API](../semantic-search/semantic-search-inference.md)**  
  Use this when you need more control over how embeddings are generated, such as choosing models or customizing your pipeline.

:::::
::::::

## Discovery and recommendations (similar items)

## Multimodal search (image, audio, video)

## Duplicate detection, fraud, and anomaly detection

> ## Enterprise and legal search across large corpora

Reviewer: "Long-term memory for LLMs" is another use case I've heard floating around, not sure how real it is though

Author: I see your point. For now, I moved the _Implementation guides and tutorials_ section to the top to highlight the OOTB semantic_text workflows, but I’m curious what others think: is this enough, or should we do something else to make `semantic_text` more prominent here?

Use this approach when building applications such as question answering over your own documents (PDFs, wikis, or tickets), internal or customer-facing knowledge assistants, chatbots powered by your own data, or customer support systems based on FAQs and ticket history. It is also commonly used for internal knowledge search across tools like Confluence, SharePoint, or Notion.

In practice, these systems are often applied to enterprise search across internal knowledge bases, customer support centers and help desks, and legal or patent search where LLMs help explore and summarize large document sets. Other common use cases include intelligence and investigation systems (for example, homeland security), academic research assistants, and tools for exploring libraries and archives.

::::::{stepper}
:::::{step} Learn about RAG in Elasticsearch

Read how retrieval-augmented generation fits into {{es}} and how it relates to search and orchestration.

- [RAG](../rag.md)

:::::

:::::{step} Choose a search strategy

> Implement retrieval that matches your content and latency needs. Use [semantic search](../semantic-search.md) workflows for managed meaning-based retrieval, [vector search](../vector.md) when you configure embeddings and fields directly, or [hybrid search](../hybrid-search.md) to combine full-text and semantic retrieval. Hybrid setups often use [reciprocal rank fusion (RRF)](elasticsearch://reference/elasticsearch/rest-apis/reciprocal-rank-fusion.md) to merge rankings.

Reviewer: again "meaning-based" isn't the distinguishing trait here: it's the field-type/workflow choice

In practice, this includes image search (for example, visual product search), audio and video search, and cross-modal scenarios such as using text queries to find images or other media.

::::::{stepper}
:::::{step} Select a multimodal embedding workflow

Use the inference APIs with a model and service that supports the modalities you need (for example, text and image inputs to a shared embedding space). Configure endpoints so documents and queries use the same task type and model; otherwise similarity scores are not comparable.

- [Elastic Inference](../../../explore-analyze/elastic-inference.md)
- [Semantic search with the inference API](../semantic-search/semantic-search-inference.md)

:::::

:::::{step} Index vectors for each media object

Store embeddings in `dense_vector` fields when vectors are produced outside a managed `semantic_text` path, and keep textual or structured fields you still filter on.

- [Bring your own dense vectors](bring-own-vectors.md)
- [Dense vector search](dense-vector.md)
- [`dense_vector` field](elasticsearch://reference/elasticsearch/mapping-reference/dense-vector.md)

:::::

:::::{step} Encode queries in the same space

> At search time, turn the user’s text or media into an embedding using the same inference endpoint and settings as at ingest (task type, model, dimensions). Your application typically sends the query payload to `POST _inference/{task_type}/{inference_id}` and uses the returned vector in the next `_search` request. If you prefer {{esql}}, the `EMBEDDING` function can produce the dense vector instead. Pass that vector into retrieval via the Search API, for example the `knn` option or a `knn` query.

Reviewer: Let's not mix ESQL and _search here

you use EMBEDDING to generate the vector and KNN to search in ESQL

See https://www.elastic.co/docs/reference/query-languages/esql/functions-operators/dense-vector-functions/embedding

Let's be clear about the difference between (_search + Query DSL) versus ESQL

::::::{stepper}
:::::{step} Normalize records before embedding

Clean and canonicalize the text or features you fingerprint (for example, normalized URLs, stripped boilerplate, hashed identifiers) so trivial differences do not fragment clusters in embedding space.

- [Ingest for search](../ingest-for-search.md)

:::::

:::::{step} Index vectors for each entity or event

Use `semantic_text` when you want {{es}} to manage inference, or `dense_vector` when you generate vectors upstream. You need one comparable vector per document or chunk you want to compare.

- [Semantic search with `semantic_text`](../semantic-search/semantic-search-semantic-text.md)
- [Bring your own dense vectors](bring-own-vectors.md)

:::::

:::::{step} Query for nearest neighbors and cutoffs

Use kNN to fetch the closest documents to a candidate and interpret similarity scores or ranks in your application, for example flagging pairs above a cosine-similarity threshold as likely duplicates, or surfacing accounts whose behavior vectors sit unusually close to known fraud exemplars.

- [kNN search](knn.md)

> - [`knn` query](elasticsearch://reference/query-languages/query-dsl/query-dsl-knn-query.md)

Reviewer: If we're going to be consistent, we need to mention both _search and ESQL

---
navigation_title: Use cases
applies_to:
  stack:
  serverless:
products:
  - id: elasticsearch
  - id: cloud-serverless
  - id: cloud-hosted
  - id: cloud-enterprise
  - id: cloud-kubernetes
  - id: elastic-stack
---

> # Vector search use cases

Reviewer: The use cases page has good structure and the stepper format works well. Main feedback is about curation. Some ideas:

**Move shared context into the intro.** The Query DSL vs ES|QL choice, the managed-vs-direct decision, and links like Search API / ES|QL for search / semantic_text / hybrid search apply to every use case. State them once at the top of the page so each use case section can focus on what's specific to it.

**Trim links per step.** Some steps have 7-9 links. That reads as a sitemap, not guidance. Aim for 1-2 per step: the link you'd _actually_ click next.

**One intro paragraph per use case, not two.** Several sections (especially RAG) have a "use this when" paragraph followed by an "in practice" paragraph that lists the same examples reworded. Pick one.

**Make each use case earn its section.** Duplicate detection and recommendations currently walk through the same generic pattern: index vectors, run kNN, filter. What makes dedup *different* from recommendations: thresholds, batch comparisons, alerting on anomalous clusters? That's the part worth covering.

> :::{tip}

Reviewer: This is a bit jarring at top of page, perhaps could move it after the first paragraph?

also the second part of the tip could potentially be woven into the prose of that first paragraph, keeping the tip small

> New to semantic search? Start with the [semantic search quickstart](../get-started/semantic-search.md), which uses the managed `semantic_text` workflow.

Reviewer: ```suggestion
New to vector search? You might want to start with the [managed `semantic_text` workflow](../get-started/semantic-search.md).
```

::::::{stepper}
:::::{step} Learn about RAG in Elasticsearch

Read how retrieval-augmented generation fits into {{es}} and how it relates to search and orchestration.

- [RAG](../rag.md)

:::::

:::::{step} Choose a search strategy

Implement retrieval that matches your content and latency needs. Use [semantic search](../semantic-search.md) when you want a managed workflow (for example the `semantic_text` field type), [vector search](../vector.md) when you configure embeddings and vector fields directly, or [hybrid search](../hybrid-search.md) to combine full-text with semantic or vector retrieval in one request. Hybrid setups often use [reciprocal rank fusion (RRF)](elasticsearch://reference/elasticsearch/rest-apis/reciprocal-rank-fusion.md) to merge rankings.

You can run retrieval using the [Search API with Query DSL](../the-search-api.md), or [{{esql}}](../esql-for-search.md).

> 

Reviewer: this is too many links I'm afraid, a user won't know where to click here

Reviewer: I think each step should have 1-2 links max to avoid overwhelm

In practice, this includes image search (for example, visual product search), audio and video search, and cross-modal scenarios such as using text queries to find images or other media.

::::::{stepper}
:::::{step} Select a multimodal embedding workflow

Use the {{infer}} APIs with a model and service that supports the modalities you need (for example, text and image inputs to a shared embedding space). Configure endpoints so documents and queries use the same task type and model; otherwise similarity scores are not comparable.

- [Elastic {{infer-cap}}](../../../explore-analyze/elastic-inference.md)
- [Semantic search with the {{infer}} API](../semantic-search/semantic-search-inference.md)

:::::

:::::{step} Index vectors for each media object

Store embeddings in `dense_vector` fields when vectors are produced outside a managed `semantic_text` path, and keep textual or structured fields you still filter on.

- [Bring your own dense vectors](bring-own-vectors.md)
- [Dense vector search](dense-vector.md)
- [`dense_vector` field](elasticsearch://reference/elasticsearch/mapping-reference/dense-vector.md)

:::::

> :::::{step} Encode queries in the same space

Reviewer: ```suggestion
:::::{step} Encode queries in the same vector space
```
